subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTestOperationListener.java

```diff
@@ -85,56 +89,53 @@ public void finished(BuildOperationDescriptor buildOperation, OperationFinishEve
+    private DefaultTestDescriptor adapt(OperationIdentifier buildOperationId, TestDescriptorInternal testDescriptor) {
+        return testDescriptor.isComposite() ? toTestDescriptorForSuite(buildOperationId, testDescriptor) : toTestDescriptorForTest(buildOperationId, testDescriptor);
     }
 
-    private DefaultTestDescriptor toTestDescriptorForSuite(TestDescriptorInternal suite) {
+    private DefaultTestDescriptor toTestDescriptorForSuite(OperationIdentifier buildOperationId, TestDescriptorInternal suite) {
         Object id = suite.getId();
         String name = suite.getName();
         String displayName = suite.toString();
         String testKind = InternalJvmTestDescriptor.KIND_SUITE;
-        String suiteName = suite.getName();
         String className = suite.getClassName();
         String methodName = null;
-        Object parentId = getParentId(suite);
-        final String testTaskPath = getTaskPath(suite);
-        return new DefaultTestDescriptor(id, name, displayName, testKind, suiteName, className, methodName, parentId, testTaskPath);
+        Object parentId = getParentId(buildOperationId, suite);
+        String testTaskPath = getTaskPath(buildOperationId);
+        return new DefaultTestDescriptor(id, name, displayName, testKind, suite.getName(), className, methodName, parentId, testTaskPath);
     }
 
-    private DefaultTestDescriptor toTestDescriptorForTest(TestDescriptorInternal test) {
+    private DefaultTestDescriptor toTestDescriptorForTest(OperationIdentifier buildOperationId, TestDescriptorInternal test) {
         Object id = test.getId();
         String name = test.getName();
         String displayName = test.toString();
         String testKind = InternalJvmTestDescriptor.KIND_ATOMIC;
-        String suiteName = null;
         String className = test.getClassName();
         String methodName = test.getName();
-        Object parentId = getParentId(test);
-        final String taskPath = getTaskPath(test);
-        return new DefaultTestDescriptor(id, name, displayName, testKind, suiteName, className, methodName, parentId, taskPath);
+        Object parentId = getParentId(buildOperationId, test);
+        String taskPath = getTaskPath(buildOperationId);
+        return new DefaultTestDescriptor(id, name, displayName, testKind, null, className, methodName, parentId, taskPath);
     }
 
-    private String getTaskPath(TestDescriptorInternal givenDescriptor) {
-        TestDescriptorInternal descriptor = givenDescriptor;
-        while (descriptor.getOwnerBuildOperationId() == null && descriptor.getParent() != null) {
-            descriptor = descriptor.getParent();
-        }
-        return runningTasks.get(descriptor.getOwnerBuildOperationId());
+    @Nullable
+    private String getTaskPath(OperationIdentifier buildOperationId) {
+        return ancestryTracker.findClosestExistingAncestor(buildOperationId, runningTasks::get)
+            .orElse(null);
```

Comment:
Should this be `orElseThrow()`? I suppose you always should find a task for the test operation.

Reply:
I'd rather not change the behavior now.

Reply:
Fine with me.